### PR TITLE
Provide add_planar_joint in model directives

### PR DESCRIPTION
**Pull request by RussTedrake**
I have many examples in the manipulation notes that restrict the dynamics to the plane. Without this, I do not have a reasonable way to construct the 2D scenes from yaml (e.g. for use in hardware_sim). It is not viable to have to add a planar joint to every model that needs to be inserted.

+@jwnimmer-tri for feature review, please?

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20122)

### CODE REVIEW

**Comment by RussTedrake** [comment]
I had considered that version, but it's very ugly for something that would be between the world and *every* object in my scene.  Ugly in the yaml and in the resulting kinematic tree.

If we really must write out an sdf, we could write out a comparable dynamics with two sliders and a pin joint?

**Comment by RussTedrake** [comment]
Also, on principle I'm not sure I like the idea of constraining ourselves to sdf's limitations. I'm not sure when that requirement came in... but maybe it's worth revisiting.

**Comment by jwnimmer-tri** [comment]
> >  ... SDFormat cannot itself express ...

My mistake; I checked http://sdformat.org/ but forgot to check the [Drake Extensions](https://drake.mit.edu/doxygen_cxx/group__multibody__parsing.html).  In the latter, we do have a `<drake:joint ...>` type for `planar`, and that syntax would be probably be OK to use as a conversion target.  Or, converting it to 3 joints (2 prismatic + 1 revolute) might also be plausible.

>  I'm not sure when that requirement came in... but maybe it's worth revisiting.

We require that `.dmd.yaml` syntax can be programmatically converted into an `*.sdf` file, so that dmd is merely sugar, not a full-blown novel semantic file format (#17223).  The hope is to eventually deprecate and remove model directives, now that SDFormat actually has a workable `<include ...>` element: [_"We expect that this mechanism will be temporary and will be removed..."_](https://github.com/RobotLocomotion/drake/blob/master/multibody/parsing/README_model_directives.md).

If we decide to keep dmd as sugar indefinitely, that's probably OK, but I do think its important to be able to "export" the dmd files into SDFormat so that users can bring their models over to other simulators, too.

> I had considered that version, but it's very ugly ...

Hmm, when I first read the top post, somehow I missed the qualifier "reasonable", and so I misunderstood that it was _impossible_ to do this, rather than just _ugly_.  I thought providing a working example would help move the discussion forward.

We seem to agree that making a planar scene is _possible_ with current tools, and you are proposing sugar for making it less ugly.  Now that we're in the land of "sugar" not just "possible", it's important to provide examples of the typical uses cases, and show some "ugly" scenarios that would benefit from the sugar.  The minimum unit test here does not help me (or others) understand why this would be so useful in the examples you must have in mind.

For example, if it is really *every* model that should be planar, then maybe we'd prefer something more like #18741 where the user configures the MbP with what should happen to "floating" bases -- keep them as 7dof floating (the default), or use 6dof floating (#14949), or 3dof planar (this PR), or weld (#18741).  That could either be directly in the MbPConfig, or a "joint finalizer" add-on that would operate on top of the plant, after all of the other models have been loaded.  If that meets the need, wouldn't that be even _better_ sugar that writing out all of the joints one by one, even with this abbreviated syntax?

Or, possibly dmd is the wrong answer for this use case in the first place?  I can imagine that an SDformat file with `<include>` statements (for the items in the scene) and planar joints (to constrain them) would probably end up looking pretty reasonable, and would not have extra weld joints.  The scenario.yaml could `add_model` on the `planar_world.sdf` scene description.

In short, showing a full demonstration or example that needs this feature would go a long way towards explaining your thought process.

**Comment by RussTedrake** [comment]
Re: use cases.  
- I tagged this PR in my update to the intro notebook in manipulation in the hopes that you'd see it.  Here's the link again.  [Use model directives planar joint](https://github.com/RussTedrake/manipulation/commit/d6c6a4aff351d240a6351bf327862595a4aaf4bb).  This example is particularly important to me, since it's many people's first experience with Drake.  If the directives there are scary, that's not ideal.
- For a more extreme example see: https://manipulation.mit.edu/clutter.html#example1 .  I could imagine continuing to populate that particular world programmatically.  But it's a proxy for the more complicated scenes that are possible.
- The [box-flipup example in gym](https://github.com/RussTedrake/manipulation/blob/4af75b1f19beff67c5423e0fb0df81114e221c8a/manipulation/envs/box_flipup.py#L65) and all of the planar dish-loading sims we did before at TRI are other good examples.

Re: "it's just sugar".  I agree it's technically possible. But I would not be happy with the current solution in the intro notebook, for instance.

**Comment by jwnimmer-tri** [comment]
I've the examples now, and so far I don't agree with the premise.  Previously, those examples were setting up the planar joint using API calls (one by one, manually adding each joint).

One question is -- can we change the code into data, i.e., can we use a file to denote the scenario, and load it?  In general, we want to be able to specify scenarios using data files, not custom code, and I agree with that as a goal.

However, I don't think this incantation is common enough to deserve bespoke _syntax_ and _semantics_ that we need to specify, teach, and maintain.

(Option 1) For fixed (i.e., unparameterized) scenarios like an arm + table + brick all hard-coded, one no-code solution is to create some scenario-specific `everything.sdf` file that `<include ...>`s the child models, and then declares the joints using SDFormat syntax.  This approach would allow for _any_ kinds of joints or customization (within SDFormat's grammar), so provides a wide avenue for users to take it further.  Once we teach the `<include ...>` statement, the rest are already things they should know about.

(Option 2) Another no-code way is what I listed above -- to use dmd, create a reusable joint using model directives, and then load and weld it in place as needed.  This allows for a list of dmd stanzas as the top-level scenario, but with the extra syntax for welding.

(Amendment 2b)  It's a very common operation to "add model and then weld it".  It _might_ be worth creating an `add_model_and_weld` directive that desugars to the two of them in series, to avoid repeating the model instance name so many times in the file.  We could survey the existing dmd uses to see if this was helpful.

(Option 3) If we really do want to expand what dmd can denote, it makes more sense to me to do that by allowing SDFormat strings inline (instead of only as file URLs), instead of customized elements.  The user could write the `<joint name="..." type="planar">` syntax directly in the dmd text, with the full power of SDFormat ready in the wings.

(Option 4) If the major complaint is "xml sucks; humans should write yaml not xml"[1], then the we would imagine creating (in principle) a full yaml transmogrification of the sdformat xml schema.  We would adopt wholesale the exact element names, attribute names, and hierarchy from SDFormat (or some blend of SDFormat and URDF).  So the syntax would be something like:

```yaml
model:
  name: foo
  link: ...
  joint:
     name: body0_to_world
     type: planar
     parent: planar_base
     child: body0::__model__
```

This could be machine-mapped into the SDFormat parser and taken from there.  The only change would be to the front-end of parsing.  The dmd2sdf output is then also straightforward.

This is a bit more to teach, but if it makes the files nicer maybe its worth it.  The teaching would be _mundane_ though, nothing bespoke.  We'd just cite the upstream SDFormat specs for the semantics of each attribute.

[1] When prototyping, I absolutely agree that YAML is better.  For publishing stable model files, SDFormat is better because has supports schema versioning and automated upgrades to newer schema versions.

(Option 5) Don't try to do it at the MbP/Parser level, rather at the Scenario level:

```
@dataclass
struct Scenario:
    directives: ...

    ...

    # When set, all floating bodies will be jointed by a planar joint to this frame name.
    planar_origin: Optional[str] = None
```

... plus a little helper function called by `ApplyScenario` to loop over the plant and add the joints.

Once we're in the regime of the user's custom scenario setup code (instead of syntax built-in to Drake), we/they can customize the semantics and defaults to their heart's content.
